### PR TITLE
[WIP] adding additional tests for notebooks

### PR DESCRIPTION
**Pull request by kalset1**
**Legal Acknowledgement**\
By contributing to this software project, I agree my contributions are submitted under the BSD license. 
I represent I am authorized to make the contributions and grant the license. 
If my employer has rights to intellectual property that includes these contributions, 
I represent that I have received permission to make contributions and grant the required license on behalf of that employer.

### CODE REVIEW

**Comment by kalset1** [comment]
mnist_convolutional and neural_network_formulations notebook have empty code cells at the bottom which cause errors

File: tests/notebooks/test_run_notebooks.py [hunk 1]
```diff
@@ -1,54 +1,295 @@
...
+# TODO: We need to try and write these tests to rely on internal consistencies and less on absolute numbers and tolerances
 
-# TODO: These will be replaced with stronger tests using testbook soon
-def _test_run_notebook(folder, notebook_fname, n_cells):
-    # change to notebook directory to allow testing
-    cwd = os.getcwd()
+
+# return testbook for given notebook
+def open_book(folder, notebook_fname, **kwargs):
+    execute = kwargs.get("execute", True)
```
Comment (carldlaird) [inline tests/notebooks/test_run_notebooks.py:15]: Can you use named arguments instead of **kwargs in these methods? Actual arguments provide better documentation for someone using the functions.

File: tests/notebooks/test_run_notebooks.py [hunk 2]
```diff
@@ -1,54 +1,295 @@
...
+def check_cell_execution(tb, notebook_fname, **kwargs):
+    injections = kwargs.get("injections", 0)
+    assert (
+        tb.code_cells_executed
+        == cell_counter(notebook_fname, only_code_cells=True) + injections
+    )
+
+
+# checks for correct type and number of layers in a model
+def check_layers(tb, activations, network):
```
Comment (carldlaird) [inline tests/notebooks/test_run_notebooks.py:31]: It might be better if the function names match more what the function does (rather than what it is used for). This function actually injects code into the notebook. Maybe "inject_activation_check"

File: tests/notebooks/test_run_notebooks.py [hunk 3]
```diff
@@ -1,54 +1,295 @@
...
+                total += 1
+        return total
+    else:
+        return len(nb.cells)
+
+
+# gets model stats for mnist notebooks
+def mnist_stats(tb, fname):
+    total_cells = cell_counter(fname)
+    tb.inject("test(model, test_loader)")
```
Comment (carldlaird) [inline tests/notebooks/test_run_notebooks.py:59]: What is this line doing? Can we add some more comments that describe what these (and other) lines are doing?

File: tests/notebooks/test_run_notebooks.py [hunk 4]
```diff
@@ -1,54 +1,295 @@
...
+    model_stats = tb.cell_output_text(total_cells)
+    model_stats = model_stats.split(" ")
+    loss = float(model_stats[4][:-1])
+    accuracy = int(model_stats[-2][:-6])
+    return (loss, accuracy)
+
+
+# neural network formulation notebook helper
+def neural_network_checker(tb, ref_string, val1, val2, tolerance):
```
Comment (carldlaird) [inline tests/notebooks/test_run_notebooks.py:68]: Same comment here. We need some comments to help understand what these are doing.

File: tests/notebooks/test_run_notebooks.py [hunk 5]
```diff
@@ -1,54 +1,295 @@
...
-    _test_run_notebook("neuralnet", "auto-thermal-reformer-relu.ipynb", 13)
+    notebook_fname = "auto-thermal-reformer-relu.ipynb"
+    book = open_book("neuralnet", notebook_fname)
+
+    with book as tb:
+        check_cell_execution(tb, notebook_fname)
+
+        # check loss of model
+        model_loss = tb.ref("nn.evaluate(x, y)")
+        assert model_loss == pytest.approx(0.000389626, abs=0.00031)
```
Comment (carldlaird) [inline tests/notebooks/test_run_notebooks.py:85]: The model loss will not be consistent enough to check this way. I think we should just check if it is at least as small as some acceptable tolerance - just so we know if the network begins training more poorly for some reason.

**Review by carldlaird — CHANGES_REQUESTED** [review]
I added some broad comments. Let's do another pass and review again.

File: tests/notebooks/test_run_notebooks.py [hunk 6]
```diff
@@ -1,54 +1,295 @@
...
+
+    with book as tb:
+        check_cell_execution(tb, notebook_fname)
+
+        # check loss of model
+        model_loss = tb.ref("nn.evaluate(x, y)")
+        assert model_loss == pytest.approx(0.00024207, abs=0.00021)
+
+        # check layers of model
+        layers = ["sigmoid", "sigmoid", "sigmoid", "sigmoid", "linear"]
```
[the remark this answers is not in the document]
Comment (kalset1) [reply tests/notebooks/test_run_notebooks.py:118]: Yes I think that makes sense. Do you think checking the layers of the imported models (the ones that are built first with pytorch and then converted using "load_onnx_neural_network_with_bounds") still makes sense? Or is that not needed
